aten/src/ATen/core/Tensor.h

```diff
@@ -1,3 +1,41 @@
 #pragma once
 
 #include <ATen/core/TensorBody.h>
+#include <c10/util/Exception.h>
+
+namespace at {
+class TORCH_API OptionalTensorRef {
+ public:
+  OptionalTensorRef() {}
+  OptionalTensorRef(const Tensor& src)
+      : ref_(c10::intrusive_ptr<TensorImpl>(
+            src.unsafeGetTensorImpl(),
+            c10::raw::DontIncreaseRefcount{})) {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(src.defined());
+  }
+
+  OptionalTensorRef(const OptionalTensorRef& rhs)
+      : ref_(c10::intrusive_ptr<TensorImpl>(
+            rhs.ref_.unsafeGetTensorImpl(),
+            c10::raw::DontIncreaseRefcount{})) {}
+
+  ~OptionalTensorRef() {
+    ref_.unsafeReleaseTensorImpl();
+  }
+
+  bool has_value() const {
+    return ref_.defined();
+  }
+
+  const Tensor& getTensorRef() const {
+    return ref_;
+  }
```

Comment:
This needs the ref qualifier as described in the RFC at https://github.com/pytorch/rfcs/pull/16/files#diff-04c8518032fa82b79c85ae7e3c210d7c9ef27bc103484cbc85728d79b42f2d6aR47 since `const Tensor&` will not induce const reference lifetime extension see https://github.com/pytorch/rfcs/pull/16/files#diff-04c8518032fa82b79c85ae7e3c210d7c9ef27bc103484cbc85728d79b42f2d6aR83-R88

Reply:
Sigh, details are hard.